DM-37584: add proposal from the Great Calibration Refactor discussions (#2)
lsst-dm/dmtn-172: index.rst

    :name: pl-jfb-ambitious-nofakes-01

The Great Calibration Refactor Proposal: February 2023
------------------------------------------------------

Goals and non-goals
"""""""""""""""""""

We're replacing essentially everything between ISR and coaddition.
ISR and coaddition will also see major changes (spurred by Calibpalooza and cell-based and chi-squared coadds, respectively) on similar timescales (we hope), but we're considering those out-of-scope and mostly orthogonal.
While we can and should implement this piecemeal when we can, we want a complete vision of what it will look like in the end, and in some cases it may be easier to replace many tasks at once.

> We lean towards merging PipelineTasks with the same dimensions that run back-to-back rather than keeping them distinct.

Reviewer: I commented on this on the main thread, but I really don't think this is the right approach. I think it would be better to spend the time on the operation side to cluster / in-memory data store. I don't really think it is appropriate at this level for task authors to design their algorithms around optimizing i/o

Author: I'll add a "when all other considerations are equal" qualifier here, but middleware is way behind on delivering on promises to pipelines already, with some major middleware deliverables still on the horizon.  Pipelines development effort and middleware development effort are not fungible, and even if they were, merging a pair of pretty-similar PipelineTasks (or, at worst, maybe three such pairs - that's really the most we might be debating here) is not a lot of work, and the same cannot be said for in-memory datastore and clustering work.

Reviewer 2: I tend to agree. The clustering+in-memory data store is pretty close now (especially now that I added support for copy-on-get) but it's all the support code that is lacking such as getting BPS to configure the in-memory datastore per-task when it sees there are intermediates in the clustering, and working out which datasets have to be copied on get and which don't have to be.

This is a bit of a shift - many smaller PipelineTasks leads to more flexibility via just pipeline definition changes, which has been very useful in prototyping, but we believe we are exiting the prototyping phase and should instead prioritize the I/O optimization and pipeline-simplicity advantages of having fewer bigger PipelineTasks.
We very much intend to continue to delegate all real algorithmic work to subtasks; it's just that each PipelineTask will tend towards having more of those.

This is probably our last best chance to get our naming conventions for dataset types and task labels under control, so we're including that in our proposal.

Conventions
"""""""""""

- Task labels are camelCase and start with a lowercase verb: "associateIsolatedStars" instead of "isolatedStarAssociation".
- Dataset type names are snake_case nouns preceded (if necessary) by adjectives: e.g. "initial_visit_summary"
- Use "source" instead of "src" or "sources".

> - Avoid "catalog" or "cat" in dataset type names; use "source", "stars", "object", or "matches" instead when appropriate.

Reviewer: I am split on this, I agree that if you are in the know, it does not add any additional information, but if you are not and just read a list of data set types, seeing `stars` does not really tell me anything. Is it a stack of postage stamps of isolated stars? Is it a data set type used in creation of synthetic stars? At least if there is a `table` or `catalog` somewhere in there it makes it a bit clearer what it is.

Author: I somewhat agree w.r.t. `stars`, but I do think `object`, `source`, and `matches` pretty clearly imply table/catalog.

> - Use some variant of "pvi" for any direct (non-difference) ``{visit, detector}`` image dataset with an image, mask, and variance plane (i.e. ``lsst.afw.image.Exposure`` or ``MaskedImage``).

Reviewer: Im not sure reusing the same `pvi` name for two different python dataset types is the best idea, it may be confusing when someone is looking to load something into a notebook.

Author: I'm not sure what you mean.  These aren't Python things, and it's no different from appending "exp" to "calexp" or "icExp" now (just better, I maintain).

> - Use some variant of "pvi" for any direct (non-difference) ``{visit, detector}`` image dataset with an image, mask, and variance plane (i.e. ``lsst.afw.image.Exposure`` or ``MaskedImage``).

Reviewer 3: Is the plan to replace these `*_pvi` datasets with the replacement for `Exposure` once it becomes available, or to make a new dataset to hold the new type? I'm not clear on whether the new object will be functionally a drop-in replacement.

Author: At this point I don't have enough confidence that the RFC-950 `Exposure`-replacement will happen to take any action on that basis.  Maybe we'll actually manage to migrate the dataset type definitions, or maybe we'll just hold our noses and add a prefix or suffix for the new ones.

- Tables that are initially per-detector that will get concatenated into per-visit tables should get a "_detector" suffix so the final thing does not need a suffix (and when we revamp later steps of the pipeline, the same for "_patch" so there's no "_tract").

> - Task labels and dataset type names are for "slots", not specific tasks or connections - usually those are 1-1, but when they are not, the label and dataset type names should remain fixed when a different task is swapped in (our "solveAstrometry" task label slot could be satisfied by either jointcal or GBDES).

Reviewer: I have long wanted this, and we tried to do it, but it is currently impossible (It will require refactoring middleware)

Author: In the sense that trying this without middleware changes makes pipelines to hard to import-and-modify to swap out tasks?  Or is there something that really does make it technically impossible?

>   Whether the same is true of the outputs is an open question we'd like to discuss.

Reviewer: This point in context with the last is confusing to me, surely the inputs to a task don't really matter (i.e. maybe one takes 3 more prerequisites) but the outputs should remain constant if you expect to substitute it in.

Author: Yes, I'm not sure what I was thinking here.  I'll delete this sentence.

- "initial" catalogs are not (ever) SDM-standardized, while final catalogs are always SDM-standardized before they are written out.

>   This may present a challenge for analysis tools that want to be able to run on either, but I'm hoping we can minimize analysis of the initial things, as we don't care how good they are in the end.

Reviewer: I think this point is just not true, and think it needs much more justification than a simple assertion.

Author: Agreed; my thinking on this has evolved while watching the PDR2 processing play out.  I'll switch this to a TODO entry, as I want to get this branch merged, but we do need to adjust the proposal to address this.

> - Do include a "final" prefix on dataset types that represent the best version of things that are nevertheless temporaries that will not be retained.

Reviewer: I agree its good to not include a "final" on the end of something, but assuming that something without a suffix is the final product is also not right. At any point another task may come along and be added to the chain, but we cannot keep updating existing dataset type names. It will really break all downstream workflows

Author: If we need to change the task that produces a final thing, to a new task that runs after the original one, the right thing to do is to change the dataset type configured for the first one, so the new task can use the final name.

If the dataset types do not have the same storage class or dimensions, that's a much bigger problem because they aren't substitutable.  I agree it could still happen, but I think it'd come with some extra considerations that make these sort of general conventions less relevant anyway.

  Do not add any prefix to dataset types that will be retained for public access.

> - Convert source and object catalogs to Parquet before ever persisting them; only use SourceCatalog/FITS to hold Footprints.

Reviewer: Why not store fooprints in a completely orthogonal way?

Author: I'd be happy to, if we had another way to do it.

>   We will retain the ``slot_*`` column names and not the underlying names they point to - downstream code should only be referring to the slot columns anyway, and in many cases we won't run multiple algorithms that could satisfy the slot.

Reviewer: Overall, I think we should re-visit the idea of slot altogether. I understand why it was introduced, but I think in practice measurements are not isometric to each other, and most downstream consumers actually care about the algorithm, such that they will always directly access that named column, or slots can never really be changed for how it will disrupt downstream consumers.

Author: I don't necessarily disagree but I'm considering that out of scope here.

>   Do not add any prefix to dataset types that will be retained for public access.

Reviewer 3: I think this should be reworded in the other order; the important thing here is that public-facing outputs are just named "X", while the things that go into those outputs are named "final_Y". I think flipping the order will make that more clear.

>   We will retain the ``slot_*`` column names and not the underlying names they point to - downstream code should only be referring to the slot columns anyway, and in many cases we won't run multiple algorithms that could satisfy the slot.

Reviewer 3: Retaining slot names in the parquet tables seems like a dangerous choice to me: there's no way to know what actually went into that slot (unlike the aliases in afw) and the name itself doesn't tell you. And that can really matter: see e.g. slot_apFlux for HSC vs. LATISS!

Author: I agree there's a provenance problem (you _should_ be able to figure it out from the configs, but that's at best a pain).  But if we drop the slot names everything downstream gets more fragile, because they have to effectively bake in upstream configuration to figure out which field name they want.  This still feels like the least bad option, though I concede it's not actually a good one.

  In addition, note that the (final) ``source`` catalogs will be SDM-standardized before being written, so the the ``slot_*`` vs. underlying name question is moot there.

Task and dataset type notes
"""""""""""""""""""""""""""

> bootstrapImage

Reviewer 3: Recommend a global change:

```suggestion
calibrateImage
```

And change the dotfile and SVG.

Author: Oops, I actually had this change sitting, unpushed, in my local git repo already.

   Initial background subtraction and detection of bright stars (galaxies are considered a nuisance here).

>    Initial versions of everything - astrometry, photometry, probably PSFs.

Reviewer 3: ```suggestion
   Initial versions of everything - background, astrometry, photometry, PSFs.
```

   Probably aperture corrections of some kind, but targeted specifically at making compensated apertures work for FGCM.
   These are all attached to its image output, ``initial_pvi``, which is a lot like today's ``calexp``.

>    This will be a fluence image with nJy pixel units, but how much of that is done by this task vs. ISR is TBD.

Reviewer 3: ```suggestion
   This will be a fluence image with nJy pixel units, but the details of the fluence (raw)->fluence (postISR) conversion will be mostly left to how we create our flat fields and how they are applied in ISR.
```

Or something like that. At least, I think this is what you were getting at here?

Author: I'll just strike the second clause entirely.  It doesn't add any information, just uncertainty.

>    Its output catalog, ``initial_source_detector``, will be converted to Parquet before it is written, but not SDM-standardized.

Reviewer 3: ```suggestion
   Its output catalog, ``initial_stars_detector``, will be converted to Parquet before it is written, but not SDM-standardized.
```

>    Footprints will be written to a separate ``SourceCatalog`` dataset, ``initial_footprints``.

Reviewer 3: ```suggestion
   Footprints will be written to a separate ``SourceCatalog`` dataset, ``initial_stars_footprints_detector``.
```

   Whether this task does multiple detection rounds (to iterate on CR detection, the source detection filter, or the detection threshold) is TBD; we'd like to minimize that.
   We would like to avoid running a deblender here (if we do, it will have to happen after we have the PSF model), but this depends on how this performs on crowded fields.

>    We are not thrilled with the name of this task and would love ideas for improvements (which should be coordinated with the names for ``bootstrapVisit``, ``finalizeImage``, ``compressImage``, and ``rebuildImage``).

Reviewer: I think this misses the legitimate use case of calexp now, that of a "good enough" image for things, even if it is not the final calibration (especially for pipelines that do not run FGCM) but I may not be following how that use case plays out in the proposed graph

Author: In pipelines that do not produce a better version of the image produced, the image produced here should have a name that does not imply that it will be superseded.  But I want to focus on the actual DRP pipeline here.

>    We are not thrilled with the name of this task and would love ideas for improvements (which should be coordinated with the names for ``bootstrapVisit``, ``finalizeImage``, ``compressImage``, and ``rebuildImage``).

Reviewer 3: This sentence can be removed: I think we're happy with `calibrateImage`.

bootstrapVisit

>    Consolidate the per-detector outputs of ``bootstrapImage`` and recover from failures on some detectors by using those that succeeded (especially for WCSs).

Reviewer 3: Maybe `mergeDetectorVisits`? It's not `calibrateVisit`, since that happens in fgcm/gbdes.

Or just `consolidateVisit`?

Author: Ah, see, this is where `bootstrap` was helpful...

But that ship has indeed sailed. I'll go with `consolidateVisit` for now.

associateIsolatedStars
   Pretty much just a renamed ``IsolatedStarAssociationTask``.

solveAstrometry
   Pretty much just a generic label for ``jointcal`` and ``GBDES``.

fgcm
   Nothing new here, except the sharding and names of the output datasets.

modelVisitBackground
   A replacement for SkyCorrectionTask, probably using PCA.
   SkyCorrectionTask will serve as a placeholder until we have something better.

>    Also takes care of subtracting the wings of bright stars.

Reviewer: How do you expect this to handle real astronomical background?

Author: We can't do background matching here, if that's what you mean; it'd have to go later, and be a part of coaddition.

I'm mostly assuming for now that we won't get background matching implemented for DR1, because nobody's working on it right now, and it's a massive complication to the whole structure here that I don't think we'll be able to nail down until/unless we have something mostly working anyway.

finalizeCharacterization
   Nothing new here except new names for the output datasets.

>    We are using "characterizations" here to mean "PSFs and aperture corrections", and are not thrilled about that, but we need something that means "PSFs and aperture corrections" that could also absorb aperture corrections being done rather differently than they are today.

Reviewer: finalizeStelarProfileCorrections?

Author: Not a bad idea, but at this point I'm just trying to get this merged and don't have energy to shepherd a naming decision.  Will punt.

>    Nothing new here except new names for the output datasets.

Reviewer 3: Since we no longer have `characterizeImage`, should we rename this? Looks like the existing task produces psf models and apcorr; if we switch to the compensated gaussian filters, we won't need apcorr any more, so could this just be `finalizePsf`?

Author: We will still need aperture corrections here, even if we can avoid needing them earlier.  I'm happy to rename new tasks (or even substantially changed ones) on this proposal, but I think this one does end up staying pretty much as it is, and I'm inclined to leave its name, too.

finalizeAstrometry
   This is a re-run of the astrometry solver with slightly different connections; I think it'll be necessary to make our (achromatic) WCSs consistent with the chromatic PSF models that we'll someday produce in ``finalizeCharacterization``.
   But this needs more thought.

updateVisitSummary
   Nothing new here (though the current task is only a few weeks old, and apparently still has some bugs).

finalizeImage
   This new task takes all of our hard-won final characterizations and calibrations of the image and produces a final ``{visit, detector}`` ``pvi`` image and full-depth, all-measurements ``final_source_detector`` catalog.

>    The latter will be SDM-standardized before it is ever written to disk.

Reviewer: I think it is always dangerous to embed final in any term, because you risk the dissertation problem of ``finalizeImageV2Updated``

Author: But these aren't files or datasets; they're dataset types, and the idea here is that those represent interfaces, not particular task outputs.  So you don't add "V2Updated", you reconfigure the original task that produced `finalizedImage` to write `penultimateImage` (or whatever) instead and let the new final task start writing `finalizedImage`.

consolidateSourceTable
   This task just concatenates the ``final_source_detector`` catalogs into a single per-visit ``source`` catalog.

compressImage

>    This task reads the ``pvi`` dataset, lossy-compresses the image and variance planes, lossless-compresses everything else (at least the mask; I don't know if compressing more than that is possible).

Reviewer: I am really not sure why you want to do lossy-compression, there are some really impressive lossless compression algorithms. And on that note, what is the average percentage difference in storage space between lossy compressed images vs lossless?

Reviewer 2: Robert Gruendl did an analysis at some point: https://dmtn-068.lsst.io

Without lossy compression the assertion was that we would not be able to archive the PVIs at all and would have to regenerate them on demand. It was a big struggle to include the PVIs in the costing model at all.

Author: Echoing Tim here: I don't think there are lossless compression of _floating point_ images that get anywhere near the compression ratios we need to make saving PVIs viable.  Lots of prior art here, but I don't know the details.

- Can we feed FGCM with compensated-filter photometry in order to run it before background estimation is complete?

- Do we need some kind of coaddition or warp-comparison to finalize our per-visit background models?  If so, will those also impact PSF modeling and/or aperture corrections?

> - Is a second round of astrometric fitting necessary to ensure PSF chromaticity is consistent with our (achromatic) WCSs?  Is it sufficient?

Reviewer: Are you trying to soak up dcr with the PSF?

Author: Yes.  Putting DCR in the WCS is both really problematic (you can't warp images to the same pixel grid for coaddition if the WCS depends on sub-band SED) and at best an approximation (it's not a shift, it's an asymmetric smearing, unless the SED is a delta function), while it can be fully corrected by having a chromatic PSF.

> - Are there small changes we could make to better share code with AP?

Reviewer 3: I believe `CalibrateImageTask` is the primary sharing point: so long as the inputs to that are fluence images with decent camera geometry models that we can fit an affine WCS on top of, the `initial_pvi` should be decent.

After that, it would be great if we could easily make a "get good coadds and enough information to vet them with DRP-like tools" pipeline (our current template generation pipeline is not as sophisticated as DRP's coadd building). Obviously, we want to share as much of the image subtraction code as possible, but I think that's going to happen anyway.

Author: I agree and I'll go ahead and delete the bullet point.

As for coadds, aside from Y1 _incremental_ templates I think it's probably a subset of a real DRP pipeline you want, not a new DRP-like pipeline.  And while it's not something we have wired together yet, I think eventually you should see that pipeline as producing a deep reference catalog for AP to consume as well as templates (including - but not only - to get object associations for alerts).  That means going a bit further than coaddition, but there's probably still enough in the DRP pipelines after that that a subset would still be worthwhile.
